src/replication/replication_control_channel_receiver.cpp

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright (C) 2008 Search Solution Corporation. All rights reserved by Search Solution.
+ *
+ *   This program is free software; you can redistribute it and/or modify
+ *   it under the terms of the GNU General Public License as published by
+ *   the Free Software Foundation; either version 2 of the License, or
+ *   (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program; if not, write to the Free Software
+ *  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+ *
+ */
+
+/*
+ * replication_control_channel_receiver.cpp - manages master control channel entries
+ */
+
+#include "replication_control_channel_receiver.hpp"
+
+#include <list>
+#include <mutex>
+
+#include "communication_channel.hpp"
+#include "stream_transfer_sender.hpp"
+#include "thread_manager.hpp"
+#include "thread_daemon.hpp"
+
+namespace cubreplication
+{
+  namespace master_ctrl
+  {
+    class chn_manager : public cubthread::task_without_context
+    {
+      public:
+	chn_manager (cubcomm::channel &&chn, cubstream::stream_ack *stream_ack);
+	void execute () override;
+
+      private:
+	cubcomm::channel m_chn;
+	cubstream::stream_ack *m_stream_ack;
+    };
+
+    std::list<cubthread::daemon *> chns;
+    std::mutex mtx;
+    cubstream::stream_ack *g_stream_ack = NULL;
+
+    chn_manager::chn_manager (cubcomm::channel &&chn, cubstream::stream_ack *stream_ack)
+      : m_chn (std::move (chn))
+      , m_stream_ack (stream_ack)
+    {
+    }
+
+    void chn_manager::execute ()
+    {
+      size_t len = sizeof (cubstream::stream_position);
+      cubstream::stream_position ack_sp;
+      css_error_code ec = m_chn.recv ((char *) &ack_sp, len);
+      if (ec != NO_ERRORS)
+	{
+	  m_chn.close_connection ();
+	  retire ();
```

Comment:
It is not safe to retire task here. Better to let the daemon retire it correctly.

In order to properly handle a connection error (e.g. free the daemon and task), it would be easier to use a separate daemon to manager all daemons/connections.